### ISSUE TITLE
Change `request` error message options type to a map of status codes

### ISSUE THREAD

**Issue by RichDom2185** [source-academy/frontend#2574]
The current implementation is not flexible enough:

https://github.com/source-academy/frontend/tree/master/src/commons/utils/RequestHelper.tsx#L21-L28

This only allows for a catch-all, and therefore, generic error message when any errors occur. However, there are quite a number of use cases where we want to show differing error messages depending on the backend status code. For example:

* 401/403 – straightforward
* 409 Conflict - resource already exists
* ... other potential error messages here

as a request can fail due to many reasons, not simply due to insufficient permissions.

Proposal:

Change `errorMessage` (when defined) from `string` to

```tsx
type RequestOptions = {
  // ...
  errorMessage?: {
    [statusCode: number]: React.ReactNode;
  };
  // ...
};
```

**Comment by RichDom2185** [comment]
@chownces Can I get your thoughts on this? Would this be something you can help to do?

**Comment by chownces** [comment]
The current implementation follows the spec that we have been using since the start of this project. Are there any new use cases that require a more detailed error handling/ response? If so, we can update this implementation to account for those.

I'm currently at summer school and won't have the bandwidth to look at this before the semester starts 😅

**Comment by RichDom2185** [comment]
So far, I have heard of one use case from team assessments:

* When saving an answer to a question, the frontend will send along a sort-of-"last retrieved" timestamp, and if the backend realises that the answer has been updated by someone else after the frontend last retrieved it, it will respond with a 409 Conflict and block the update operation to prevent people from accidentally overwriting others' work. Can I confirm this? @CheongYeeMing
  * Of course, there's more than one reason saving an answer can be rejected, e.g. the assessment has closed

Having said that, this is just a minor UX thing, so there's no urgency to this.

**Comment by chownces** [comment]
> So far, I have heard of one use case from team assessments:
> 
> * When saving an answer to a question, the frontend will send along a sort-of-"last retrieved" timestamp, and if the backend realises that the answer has been updated by someone else after the frontend last retrieved it, it will respond with a 409 Conflict and block the update operation to prevent people from accidentally overwriting others' work. Can I confirm this? @CheongYeeMing
>   
>   * Of course, there's more than one reason saving an answer can be rejected, e.g. the assessment has closed
> 
> Having said that, this is just a minor UX thing, so there's no urgency to this.

Yep, we should take a look at this in the near future.

**Comment by CheongYeeMing** [comment]
> * When saving an answer to a question, the frontend will send along a sort-of-"last retrieved" timestamp, and if the backend realises that the answer has been updated by someone else after the frontend last retrieved it, it will respond with a 409 Conflict and block the update operation to prevent people from accidentally overwriting others' work. Can I confirm this? @CheongYeeMing

Yes, the frontend will send a "last modified" timestamp. If anyone has overwritten the last modified copy of the answer, backend will respond with 409 conflict, they will then be notified in front-end when trying to overwrite the existing updated copy.
